sklearn/linear_model/_glm/distribution.py

```diff
@@ -0,0 +1,382 @@
+        Returns
+        -------
+        deviance: array, shape (n_samples,)
+            Computed deviance
+        """
+        pass  # pragma: no cover
+
+    def unit_deviance_derivative(self, y, mu):
+        """Compute the derivative of the unit deviance w.r.t. mu.
+
+        The derivative of the unit deviance is given by
+        ∂ d(y,μ)/(∂ μ) = -2(y-μ)/v(μ) with unit variance v(μ).
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+
+        mu : array, shape (n_samples,)
+            Predicted mean.
+        """
+        return -2 * (y - mu) / self.unit_variance(mu)
+
+    def deviance(self, y, mu, weights=1):
+        """Compute the deviance.
+
+        The deviance is a weighted sum of the per sample unit deviances,
+        D = sum_i s_i * d(y_i,μ_i)
+        with weights s_i and unit deviance d(y,μ).
+        In terms of the log-likelihood it is
+        D = -2φ * (loglike(y,μ,φ/s) - loglike(y,y,φ/s)).
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+
+        mu : array, shape (n_samples,)
+            Predicted mean.
+
+        weights : array, shape (n_samples,) (default=1)
+            Weights or exposure to which variance is inverse proportional.
+        """
+        return np.sum(weights * self.unit_deviance(y, mu))
+
+    def deviance_derivative(self, y, mu, weights=1):
+        """Compute the derivative of the deviance w.r.t. mu.
+
+        It gives ∂ D(y, μ; weights)/(∂ μ).
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+
+        mu : array, shape (n_samples,)
+            Predicted mean.
+
+        weights : array, shape (n_samples,) (default=1)
+            Weights or exposure to which variance is inverse proportional.
+        """
+        return weights * self.unit_deviance_derivative(y, mu)
+
+    def _mu_deviance_derivative(self, coef, X, y, weights, link):
+        """Compute mu and the derivative of the deviance w.r.t coef."""
+        lin_pred = _safe_lin_pred(X, coef)
+        mu = link.inverse(lin_pred)
+        d1 = link.inverse_derivative(lin_pred)
+        temp = d1 * self.deviance_derivative(y, mu, weights)
+        if coef.size == X.shape[1] + 1:
+            devp = np.concatenate(([temp.sum()], temp @ X))
+        else:
+            devp = temp @ X  # same as X.T @ temp
+        return mu, devp
+
+
+class TweedieDistribution(ExponentialDispersionModel):
+    """A class for the Tweedie distribution.
+
+    A Tweedie distribution with mean μ=E[Y] is uniquely defined by it's
+    mean-variance relationship Var[Y] ∝ μ^power.
+
+    Special cases are:
+
+    ===== ================
+    Power Distribution
+    ===== ================
+    0     Normal
+    1     Poisson
+    (1,2) Compound Poisson
+    2     Gamma
+    3     Inverse Gaussian
+
+    Parameters
+    ----------
+    power : float (default=0)
```

Comment:
We changed here from `p` to `power`. But the metric `mean_tweedie_deviance`, https://github.com/scikit-learn/scikit-learn/blob/db23c4ece02b55c51987d78420c7616917f82031/sklearn/metrics/regression.py#L629 still uses `p`. Since the metric internally uses the class `TweedieDistribution`, error messages use `power` instead of `p`.
Should we also change `mean_tweedie_deviance` from `p` to `power`?

Reply:
I agree we should use either `p` or `power` consistently. We can still change `mean_tweedie_deviance` signature since a release hasn't happened yet.

Are we sure we want to change `p` -> `power` everywhere though (and not the opposite)? Another metric we have, Minkownski distance, uses the `p` for power, and we could also argue for consistency with it as well. Any opinions on that?

Reply:
Most important is consistency. I'm fine with `p` and good documentation for it.